volatility/framework/plugins/windows/hashdump.py

```diff
@@ -0,0 +1,346 @@
+        if enc_hash == "":
+            return ""
+        (des_k1,des_k2) = self.sid_to_key(rid)
+        d1 = DES.new(des_k1.encode('latin1'), DES.MODE_ECB)
+        d2 = DES.new(des_k2.encode('latin1'), DES.MODE_ECB)
+        cipher = AES.new(hbootkey[:16], AES.MODE_CBC, salt)
+        obfkey = cipher.decrypt(enc_hash)
+        return d1.decrypt(obfkey[:8]) + d2.decrypt(obfkey[8:16])
+        
+    def get_user_hashes(self, user, samhive, hbootkey):
+        ## Will sometimes find extra user with rid = NAMES, returns empty strings right now
+        try:
+            rid = int(str(user.get_name()), 16)
+        except ValueError:
+            return None
+        V = None
+        for v in user.get_values():
+            if v.get_name() == 'V':
+                V = samhive.read(v.Data+4, v.DataLength)
+        if not V:
+            return None
+        
+        lm_offset = unpack("<L", V[0x9c:0xa0])[0] + 0xCC
+        lm_len = unpack("<L", V[0xa0:0xa4])[0]
+        nt_offset = unpack("<L", V[0xa8:0xac])[0] + 0xCC
+        nt_len = unpack("<L", V[0xac:0xb0])[0]
+
+
+        lm_revision = V[lm_offset + 2:lm_offset + 3]
+        if lm_revision == b'\x01':
+            lm_exists = True if lm_len == 20 else False
+            enc_lm_hash = V[lm_offset + 0x04:lm_offset + 0x14] if lm_exists else ""
+            lmhash = self.decrypt_single_hash(rid, hbootkey, enc_lm_hash, self.almpassword)
+        elif lm_revision == b'\x02':
+            lm_exists = True if lm_len == 56 else False
+            lm_salt = V[hash_offset+4:hash_offset+20] if lm_exists else ""
+            enc_lm_hash = V[hash_offset+20:hash_offset+52] if lm_exists else ""
+            lmhash = self.decrypt_single_salted_hash(rid, hbootkey, enc_lm_hash, self.almpassword, lm_salt)
+
+        # NT hash decryption
+        nt_len = unpack("<L", V[0xac:0xb0])[0]
+
+        nt_revision = V[nt_offset + 2:nt_offset + 3]
+        if nt_revision == b'\x01':
+            nt_exists = True if nt_len == 20 else False
+            enc_nt_hash = V[nt_offset+4:nt_offset+20] if nt_exists else ""
+            nthash = self.decrypt_single_hash(rid, hbootkey, enc_nt_hash, self.antpassword)
+        elif nt_revision == b'\x02':
+            nt_exists = True if nt_len == 56 else False
+            nt_salt = V[nt_offset+8:nt_offset+24] if nt_exists else ""
+            enc_nt_hash = V[nt_offset+24:nt_offset+56] if nt_exists else ""
+            nthash = self.decrypt_single_salted_hash(rid, hbootkey, enc_nt_hash, self.antpassword, nt_salt)
+        return lmhash, nthash
+
+    def sid_to_key(self, sid):
+
+        s1 = ""
+        s1 += chr(sid & 0xFF)
+        s1 += chr((sid >> 8) & 0xFF)
+        s1 += chr((sid >> 16) & 0xFF)
+        s1 += chr((sid >> 24) & 0xFF)
+        s1 += s1[0]
+        s1 += s1[1]
+        s1 += s1[2]
+        s2 = s1[3] + s1[0] + s1[1] + s1[2]
+        s2 += s2[0] + s2[1] + s2[2]
+        return self.str_to_key(s1), self.str_to_key(s2)
+
+    def str_to_key(self, s):
```

Comment:
Again, documentation to help figure out what all this is supposed to be do (is it part of an encryption algorithm, is it just rotating bytes?  Since our code is often referred to by people to find out how to do things themselves, it's quite important to us that the code is clear and well documentation so that those reading it can understand it.